A3-channel-tracing.md

```diff
@@ -0,0 +1,204 @@
+
+## Proposal
+
+To implement channel tracing, we will introduce a dedicated, in-memory tracing object that is owned by each channel and subchannel. We will add a new function to the grpc channel api that allows library users to retrieve all of the trace for the channel as a JSON or protobuf object. This returned trace will eventually be hooked into a UI, or exported through the ongoing grpcz project, where it could be used for debugging channel issues. The tracer will be enabled or disabled via a new channel argument, GRPC_ARG_CHANNEL_TRACING.
+
+### Format of Exported Data
+
+The trace data will be exported as JSON from the C-core. However, it will be JSON that is of the correct form to be transformed to protobuf and back again. The JSON schema will be as follows (note that all time strings will follow [RFC 3339](https://tools.ietf.org/html/rfc3339)):
+
+```
+{
+  "channelData": {
+    "numNodesLogged": number,
+    "startTime": timestamp string,
+    "nodes": [
+      {
+        "data": data,
+        "error": string,
+        "time": timestamp string,
+        // can only be one of "GRPC_CHANNEL_INIT", etc etc 
+        "state": enum string,
+        // id of referenced subchannel
+        "subchannelId": number
+      },
+    ]
+  },
+  "numSubchannelsSeen": number,
+  "subchannelData": [
+    {
+      "subchannelId": number,
+      "numNodesLogged": number,
+      "startTime": timestamp string,
+      "nodes": [
+        {
+          "data": data,
+          "error": string,
+          "time": timestamp string,
+          "state": enum string,
+        },
+      ]
+    },
+  ]
+}
+```
+
+This matches the protobuf form of the exported trace:
+
+```protobuf
+// one single node of trace data
+message TraceNode {
+  // static string describing what event the channel went through
+  string data = 1;
+  // error assosiated with the channel event
+  string error = 2;
+  // time the trace was logged
+  google.protobuf.Timestamp time = 3;
+  // unique id of subchannel associated with this trace
+  // only used for trace present in the parent
+  int64 subchannel_id = 4;
+
+  enum ChannelState {
+    GRPC_CHANNEL_UNKNOWN = 0;
+    // channel has just been initialized
+    GRPC_CHANNEL_INIT = 1;
+    // channel is idle
+    GRPC_CHANNEL_IDLE = 2;
+    // channel is connecting
+    GRPC_CHANNEL_CONNECTING = 3;
+    // channel is ready for work
+    GRPC_CHANNEL_READY = 4;
+    // channel has seen a failure but expects to recover
+    GRPC_CHANNEL_TRANSIENT_FAILURE = 5;
+    // channel has seen a failure that it cannot recover from
+    GRPC_CHANNEL_SHUTDOWN = 6;
+  };
+  // state the channel was in when the trace was logged
+  ChannelState state = 5;
+};
+
+// trace data for one channel or subchannel
+message TraceData {
+  // unique id of subchannel associated with this trace data
+  // set to zero if this is a channel's trace data
+  int64 subchannel_id = 1;
```

Comment:
How about generalizing this so that it's not specific to subchannels (i.e., even the parent channel should have an ID)?  In the parent, this can be set to the server URI; in the subchannels, it can be set to the URI of the server address to which the subchannel is connected.